src/ocean/util/UnixSocketCommandInterface.d

```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+
+    Utility for handling requests via a unix socket.
+
+    Commands can be registered to the extension and when the command is received
+    the provided delegate will be called. Any provided arguments will be split
+    by the space character and provided as an array.
+
+    Copyright:
+        Copyright (c) 2009-2018 Sociomantic Labs GmbH.
+        All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE_BOOST.txt for details.
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.util.UnixSocketCommandInterface;
+
+import ocean.transition;
+import ocean.core.Verify;
+
+import ocean.core.Enforce;
+import ocean.io.select.EpollSelectDispatcher;
+
+/// ditto
+public class UnixSocketCommandInterface
+{
+    import core.sys.posix.sys.stat;
+
+    import ocean.core.Buffer;
+    import ocean.core.array.Transformation: filter, split;
+    import ocean.io.select.EpollSelectDispatcher;
+    import ocean.net.server.unix.UnixListener;
+    import ocean.text.convert.Integer;
+    import ocean.text.util.StringC;
+
+    /// Unix listener with custom command handling.
+    private UnixSocketListener!(typeof(this)) unix_listener;
+
+    /// Stores the command arguments split by " ";
+    private cstring[] args_buf;
+
+    /// Alias for our handler delegate.
+    public alias void delegate ( cstring[], void delegate (cstring) ) Handler;
+
+    /// Our registered map of handlers by command.
+    private Handler[istring] handlers;
+
+    /// Path to create the unix socket.
+    private cstring path;
+
+    /// Mode to apply to the unix socket after binding
+    private int mode = -1;
+
+
+    /***************************************************************************
+
+        Binds to the path and listens for the new commands
+
+        Params:
+            epoll = Epoll instance.
+            path = path to bind the socket to
+            mode = mode to apply to the socket
+
+    ***************************************************************************/
+
+    public void listen ( EpollSelectDispatcher epoll,
```

Comment:
I feel slightly uncomfortable about this class being basically about string parsing and handler lookup, but then this one method (ok, `shutdown` too) involves networking. Any thoughts?

Reply:
So, the exiting API of the `UnixSocketListener` is that it accepts some object that has a `handle` method and it calls that method to handle the request. So, that bit of the class (along with the `addHandler`, etc) can surely be moved to a separate struct, and from there we can go two ways:

1. Either we still maintain `UnixSocketCommandInterface` as a convenient way to create a socket and handle the commands (we just instantiate), or
2. We just leave the handler bits. However, then user falls into a trap where they need to figure out, honestly, not so trivial `UnixSocketListener`'s API and how to plug these bits together. 

Over these two, I would prefer 1, but I honestly don't see downsides to keep everything in a single struct. However, I was wrong before about the modularization (see the reason for this PR as a reference), so I am keen to change things.

Reply:
I've implemented the approach described in 1 - I've split the handler's registry functionality into a separate class, and created the handy helper to combine the `UnixListener` and `CommandsRegistry` API.